src/vs/workbench/contrib/notebook/browser/notebookCellStatusBarServiceImpl.ts

```diff
@@ -3,51 +3,46 @@
 import { URI } from 'vs/base/common/uri';
 import { INotebookCellStatusBarService } from 'vs/workbench/contrib/notebook/common/notebookCellStatusBarService';
-import { INotebookCellStatusBarEntry } from 'vs/workbench/contrib/notebook/common/notebookCommon';
+import { INotebookCellStatusBarItemList, INotebookCellStatusBarItemProvider, notebookDocumentFilterMatch } from 'vs/workbench/contrib/notebook/common/notebookCommon';
 
 export class NotebookCellStatusBarService extends Disposable implements INotebookCellStatusBarService {
 
-	private _onDidChangeEntriesForCell = new Emitter<URI>();
-	readonly onDidChangeEntriesForCell: Event<URI> = this._onDidChangeEntriesForCell.event;
+	private _onDidChangeProviders = new Emitter<void>();
+	readonly onDidChangeProviders: Event<void> = this._onDidChangeProviders.event;
 
-	private _entries = new ResourceMap<Set<INotebookCellStatusBarEntry>>();
+	private _onDidChangeItems = new Emitter<void>();
+	readonly onDidChangeItems: Event<void> = this._onDidChangeItems.event;
 
-	private removeEntry(entry: INotebookCellStatusBarEntry) {
-		const existingEntries = this._entries.get(entry.cellResource);
-		if (existingEntries) {
-			existingEntries.delete(entry);
-			if (!existingEntries.size) {
-				this._entries.delete(entry.cellResource);
-			}
-		}
+	private _providers: INotebookCellStatusBarItemProvider[] = [];
 
-		this._onDidChangeEntriesForCell.fire(entry.cellResource);
+	constructor() {
+		super();
 	}
 
-	addEntry(entry: INotebookCellStatusBarEntry): IDisposable {
-		const existingEntries = this._entries.get(entry.cellResource) ?? new Set();
-		existingEntries.add(entry);
-		this._entries.set(entry.cellResource, existingEntries);
+	registerCellStatusBarItemProvider(provider: INotebookCellStatusBarItemProvider): IDisposable {
+		this._providers.push(provider);
+		let changeListener: IDisposable | undefined;
+		if (provider.onDidChangeStatusBarItems) {
+			changeListener = provider.onDidChangeStatusBarItems(() => this._onDidChangeItems.fire());
+		}
 
-		this._onDidChangeEntriesForCell.fire(entry.cellResource);
+		this._onDidChangeProviders.fire();
 
-		return {
-			dispose: () => {
-				this.removeEntry(entry);
-			}
-		};
+		return toDisposable(() => {
+			changeListener?.dispose();
+			const idx = this._providers.findIndex(p => p === provider);
+			this._providers.splice(idx, 1);
+		});
 	}
 
-	getEntries(cell: URI): INotebookCellStatusBarEntry[] {
-		const existingEntries = this._entries.get(cell);
-		return existingEntries ?
-			Array.from(existingEntries.values()) :
-			[];
+	async getStatusBarItemsForCell(docUri: URI, cellIndex: number, viewType: string, token: CancellationToken): Promise<INotebookCellStatusBarItemList[]> {
+		const providers = this._providers.filter(p => notebookDocumentFilterMatch(p.selector, viewType, docUri));
+		return await Promise.all(providers.map(p => p.provideCellStatusBarItems(docUri, cellIndex, token)));
```

Comment:
I'd be more graceful and have try-catch for `provide`-invocations. With that you prevent total failure in the presence of one misbehaving provider